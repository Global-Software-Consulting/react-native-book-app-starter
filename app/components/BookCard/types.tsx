import { Books } from 'screens/BookDetail/types';

export interface Props {
    id: number;
    hideIcon?: boolean;
    url?: string;
    bookTitle?: string;
    styleSelect: 'General' | 'Custom' | 'Large' | 'ExtraLarge';
    authorName?: string;
    isFavorite?: boolean;
    book?: Array<Books> | Books;
}

export interface IData {
    id?: number;
    bookId: number;
    averageRating: number;
    title: string;
    numberOfPages: string | number;
    shortSummary: string;
    book: {
        title: string;
        id: number;
    };
}
