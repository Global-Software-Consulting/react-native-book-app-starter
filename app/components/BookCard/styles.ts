import { Dimensions, StyleSheet } from 'react-native';
import DeviceInfo from 'react-native-device-info';
import { useTheme } from 'react-native-paper';
import { getPercentageHeight, getPercentageWidth } from 'utils/dimentionUtil';
export const useStyles = () => {
    const width = getPercentageWidth();
    const height = getPercentageHeight();
    const windowWidth = Dimensions.get('window').width;
    const windowHeight = Dimensions.get('window').height;
    const theme = useTheme();
    const styles = StyleSheet.create({
        container: {
            flex: 1,
            paddingVertical: 45,
            paddingHorizontal: 30,
            backgroundColor: 'white',
        },
        name: {
            fontSize: height('3%'),
            fontWeight: 'bold',
            fontFamily: 'Avenir-Medium',
            alignItems: 'center',
        },
        tagLine: {
            fontSize: height('2%'),
            fontFamily: 'Avenir-Medium',
            alignItems: 'center',
        },
        listCaption: {
            marginTop: 15,
            fontSize: height('2%'),
            fontFamily: 'Avenir-Medium',
            alignItems: 'center',
            fontWeight: 'bold',
        },
        button: {
            backgroundColor: '#61e3a5',
            padding: 10,
            borderRadius: 10,
            margin: 10,
        },
        horizontalRuler: {
            borderBottomColor: '#DCDCDC',
            borderBottomWidth: 1,
            marginTop: 30,
        },
        flatList: {
            flexGrow: 0,
            marginTop: 10,
        },

        books: {
            height: height('2%'),
            width: width('1%'),
            borderRadius: 4,
        },

        bookTrending: {
            height: windowHeight > windowWidth ? height('25%') : height('50%'),
            width: windowHeight > windowWidth ? width('37%') : width('15%'),
            borderRadius: 15,

        },

        bookGeneral: {
            height: windowHeight > windowWidth ? height('25%') : height('40%'),
            width: windowHeight > windowWidth ? width('30%') : width('14%'),
            borderRadius: 15,
        },

        bookLarge: {
            height: windowHeight > windowWidth ? height('30%') : height('50%'),
            width: windowHeight > windowWidth ? width('40%') : width('20%'),
            borderRadius: 15,
            margin: 10,

        },
        bookLargeShimmer: {
            height: windowHeight > windowWidth ? height('30%') : height('50%'),
            width: windowHeight > windowWidth ? width('40%') : width('20%'),
            borderRadius: 15,
            margin: 10,
            backgroundColor: '#E7E5E7',

        },

        bookExtraLarge: {
            height: windowHeight > windowWidth ? height('30%') : height('60%'),
            width: width('65%'),
            borderRadius: 20,
            margin: 5,
            alignSelf: 'center',

        },

        bookTrendingShimmer: {
            height: windowHeight > windowWidth ? height('25%') : height('50%'),
            width: windowHeight > windowWidth ? width('37%') : width('15%'),
            borderRadius: 15,
            margin: 5,

            backgroundColor: '#E7E5E7',

        },
        bookTrendingExtraLargeShimmer: {
            height: windowHeight > windowWidth ? height('25%') : height('60%'),
            width: width('65%'),
            borderRadius: 20,
            margin: 5,
            alignSelf: 'center',

        },

        textTitleEnlarged: {
            fontSize: windowHeight > windowWidth ? width('5%') : height('5%'),
            maxWidth: width('60%'),
            fontWeight: '500',
            alignSelf: 'center',
            textAlign: 'center',
        },
        authorTitleEnlarged: {
            maxWidth: width('55%'),
            color: 'red',
            fontWeight: '300',
            alignSelf: 'center',
            fontSize: windowHeight > windowWidth ? width('4%') : height('4%'),
        },

        bookGeneralShimmer: {
            height: windowHeight > windowWidth ? height('25%') : height('40%'),
            width: windowHeight > windowWidth ? width('30%') : width('14%'),
            borderRadius: 15,
            margin: 5,
            backgroundColor: '#E7E5E7',
        },

        bookViewIOS: {
            shadowColor: theme.colors.highlight,
            shadowOffset: { width: 0, height: 2 },
            shadowOpacity: 3,
            shadowRadius: 2,
            elevation:2,    
            backgroundColor: 'transparent',
            margin: 2,
            borderRadius: 20,
            marginRight: 3,
        },
        bookViewAndroid: {
            backgroundColor: 'transparent',
            margin: 2,
            borderRadius: 20,
            marginRight: 3,
        },


        heartIconTrending: {
            position: 'absolute',
            marginVertical: windowHeight > windowWidth ? height('21%') : height('40%'),
            marginHorizontal: DeviceInfo.isTablet()
                ? width('9%')
                : windowHeight > windowWidth
                ? width('4%')
                : width('1%'),
        },
        heartIconLarge: {
            position: 'absolute',
            marginVertical: windowHeight > windowWidth ? height('26%') : height('42%'),
            marginHorizontal: DeviceInfo.isTablet() ? width('9%') : width('5%'),
        },
        textTitle: {
            maxWidth: windowHeight > windowWidth ? width('30%') : width('15%'),
            fontWeight: 'bold',
            alignSelf: 'center',
            textAlign: 'center',
        },
        textLargeTitle: {
            maxWidth: windowHeight > windowWidth ? width('30%') : width('10%'),
            fontWeight: 'bold',
            alignSelf: 'center',
            textAlign: 'center',
        },
        authorTitle: {
            maxWidth: width('30%'),
            fontWeight: '200',
            alignSelf: 'center',
            textAlign: 'center',
        },
        textTitleShimmer: {
            marginTop: 5,
            maxWidth: windowHeight > windowWidth ? width('30%') : width('15%'),
            fontWeight: 'bold',
            alignSelf: 'center',
            textAlign: 'center',
        },
        heartIconGeneral: {
            position: 'absolute',
            marginVertical: windowHeight > windowWidth ? height('19%') : height('32%'),
            marginHorizontal: DeviceInfo.isTablet()
                ? width('7%')
                : windowHeight > windowWidth
                ? width('2%')
                : width('2%'),
        },
    });
    return styles;
};
