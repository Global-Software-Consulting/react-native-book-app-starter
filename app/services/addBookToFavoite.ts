import ApiConfig from '../config/api-config';
import Api from './client';

export default function addBookToFavorites(id: number) {
    return Api(ApiConfig.BASE_URL + ApiConfig.ADDTOFAVORITES + id, null, 'POST');
}
