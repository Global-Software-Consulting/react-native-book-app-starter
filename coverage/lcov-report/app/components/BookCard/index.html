
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for app/components/BookCard</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../index.html">All files</a> app/components/BookCard</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">53.65% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>22/41</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">48.62% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>53/109</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">45.45% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>5/11</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">53.84% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>21/39</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="BookCard.tsx"><a href="BookCard.tsx.html">BookCard.tsx</a></td>
	<td data-value="36.66" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 36%"></div><div class="cover-empty" style="width: 64%"></div></div>
	</td>
	<td data-value="36.66" class="pct low">36.66%</td>
	<td data-value="30" class="abs low">11/30</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="39" class="abs low">13/39</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="9" class="abs low">3/9</td>
	<td data-value="35.71" class="pct low">35.71%</td>
	<td data-value="28" class="abs low">10/28</td>
	</tr>

<tr>
	<td class="file high" data-value="BookCardShimmer.tsx"><a href="BookCardShimmer.tsx.html">BookCardShimmer.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="81.25" class="pct high">81.25%</td>
	<td data-value="16" class="abs high">13/16</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	</tr>

<tr>
	<td class="file high" data-value="styles.ts"><a href="styles.ts.html">styles.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="54" class="abs medium">27/54</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank">istanbul</a>
                at Wed Nov 10 2021 14:36:08 GMT+0500 (Pakistan Standard Time)
            </div>
        </div>
        <script src="../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../sorter.js"></script>
        <script src="../../../block-navigation.js"></script>
    </body>
</html>
    