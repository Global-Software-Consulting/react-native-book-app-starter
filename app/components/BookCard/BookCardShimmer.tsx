import React from 'react';
import { View } from 'react-native';
//importing style
import { useStyles } from './styles';

interface Props {
    styleSelect: 'General' | 'Custom' | 'Large';
}
const BookCardShimmer: React.FC<Props> = ({ styleSelect }) => {
    const styles = useStyles();
    return (
        <View style={styles.bookViewIOS}>
            <View
                style={
                    styleSelect === 'General'
                        ? styles.bookGeneralShimmer
                        : styleSelect === 'Custom'
                        ? styles.bookTrendingShimmer
                        : styleSelect === 'Large'
                        ? styles.bookLargeShimmer
                        : styles.bookTrendingExtraLargeShimmer
                }
            />
            <View
                style={
                    styleSelect === 'General' || styleSelect === 'Custom' || styleSelect === 'Large'
                        ? styles.textTitleShimmer
                        : styles.textTitleShimmer
                }
            />
            <View
                style={
                    styleSelect === 'General' || styleSelect === 'Custom' || styleSelect === 'Large'
                        ? styles.textTitleShimmer
                        : styles.textTitleShimmer
                }
            />
        </View>
    );
};

export default BookCardShimmer;
