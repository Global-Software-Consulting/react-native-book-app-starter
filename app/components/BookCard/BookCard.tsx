//to update heart icon on focus
import { useIsFocused } from '@react-navigation/native';
import { FavoriteBook } from 'models/reducers/appReducers';
import { ReducerState } from 'models/reducers/index';
import React, { useEffect, useState } from 'react';
import { Platform, View } from 'react-native';
//image with placeholder
import FastImage from 'react-native-fast-image';
import { Text } from 'react-native-paper';
//verctor icons
import Icon from 'react-native-vector-icons/FontAwesome';
//for responsive screen
import { useDispatch, useSelector } from 'react-redux';
import addBookToFavoite from 'services/addBookToFavoite';
import removeBookFromFavoite from 'services/removeBookFromFavoite';
import * as appActions from 'store/actions/appActions';
//importing style
import { useStyles } from './styles';
import { IData, Props } from './types';

const BookCard: React.FC<Props> = ({
    id,
    url,
    styleSelect,
    bookTitle,
    hideIcon,
    authorName,
    book,
}) => {
    const favoriteBooks = useSelector((state: ReducerState) => state.appReducer.favorite);
    const isFocused = useIsFocused();
    const [isFavorite, setIsFavorite] = useState<boolean>(false);
    const dispatch = useDispatch();
    const styles = useStyles();
    useEffect(() => {
        favoriteBooks?.findIndex((value: IData) => {
            if (value?.bookId === id) {
                setIsFavorite(true);
            }
        });
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [isFocused]);

    const apiAddFavorite = async () => {
        if (isFavorite) {
            //filtering out new data after deletion

            const newData: Array<FavoriteBook> = favoriteBooks.filter((item) => item.bookId !== id);

            //removing the red heart icon
            setIsFavorite(false);

            //passing on the new data to be set as favorite
            dispatch(appActions.setNewFavorites(newData));

            //calling the api for removal of data
            removeBookFromFavoite(id).then((response) => {
                if (response && response?.status === 'success') {
                    return response;
                } else {
                    setIsFavorite(true);
                }
                return response;
            });
        } else {
            setIsFavorite(true);

            addBookToFavoite(id)
                .then((response) => {
                    if (response && response.status === 'success') {
                        return response;
                    } else {
                        setIsFavorite(false);
                    }
                    return response;
                })
                .catch(() => {
                    setIsFavorite(false);
                });
        }
    };

    return (
        <View style={Platform.OS === 'android' ? styles.bookViewAndroid : styles.bookViewIOS}>
            <FastImage
                source={{ uri: url, priority: FastImage.priority.normal }}
                resizeMode="contain"
                style={
                    styleSelect === 'General'
                        ? styles.bookGeneral
                        : styleSelect === 'Custom'
                        ? styles.bookTrending
                        : styleSelect === 'Large'
                        ? styles.bookLarge
                        : styles.bookExtraLarge
                }
            />
            {!hideIcon && (
                <Icon
                    name="heart"
                    size={25}
                    style={
                        styleSelect === 'General'
                            ? styles.heartIconGeneral
                            : styleSelect === 'Custom'
                            ? styles.heartIconTrending
                            : styles.heartIconLarge
                    }
                    color={isFavorite ? 'red' : 'white'}
                    onPress={apiAddFavorite}
                />
            )}
            <Text
                style={
                    styleSelect === 'General' || styleSelect === 'Custom'
                        ? styles.textTitle
                        : styleSelect === 'Large'
                        ? styles.textLargeTitle
                        : styles.textTitleEnlarged
                }>
                {bookTitle}
            </Text>
            <Text
                style={
                    styleSelect === 'General' || styleSelect === 'Custom' || styleSelect === 'Large'
                        ? styles.authorTitle
                        : styles.authorTitleEnlarged
                }>
                {authorName}
            </Text>
        </View>
    );
};

export default BookCard;
