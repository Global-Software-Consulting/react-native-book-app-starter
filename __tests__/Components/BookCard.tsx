import { act, fireEvent, render } from '@testing-library/react-native'
import React from 'react'
import * as redux from 'react-redux'
import BookCard from './../../app/components/BookCard/BookCard'
import fetch, { Response } from 'node-fetch'
import addBookToFavoite from './../../app/services/addBookToFavoite'
jest.mock('react-native-modal', () => {
    return {
        __esModule: true,
        A: true,
        Modal: jest.fn(),
        default: 'mockedDefaultExport',
    }
})
jest.mock('./../../workaround/useEffect', () => {
    return { useEffect: require('react').useLayoutEffect }
})
jest.mock('react-native-paper', () => {
    return {
        __esModule: true,
        A: true,
        Text: jest.fn(),
        default: 'mockedDefaultExport',
    }
})
jest.mock('react-native-device-info', () => {
    return {
        DeviceInfo: { isTablet: () => true },
    }
})
jest.mock('react-native-paper', () => {
    return {
        Text: jest.fn(),
        useTheme: () => {
            return {
                theme: { colors: { highlight: jest.fn() } },
            }
        },
    }
})

jest.mock('react-native-device-info', () => {
    return {
        isTablet: jest.fn(),
    }
})

jest.mock('@react-navigation/native', () => {
    return {
        __esModule: true,
        A: true,
        useIsFocused: jest.fn(() => {
            isFocused = false
        }),
        default: 'mockedDefaultExport',
    }
})
jest.mock('react-native-fast-image', () => {
    return {
        __esModule: true,
        A: true,
        FastImage: jest.fn(),
        priority: {
            normal: 'normal',
        },
        default: 'mockedDefaultExport',
    }
})

jest.mock('react-redux', () => {
    return {
        __esModule: true,
        A: true,
        useDispatch: jest.fn(),
        useSelector: jest.fn(),
        default: 'mockedDefaultExport',
    }
})
jest.mock('React', () => ({
    ...jest.requireActual('React'),
    useEffect: jest.fn(),
}))
describe('Component testing', () => {
    afterEach(() => {
        jest.clearAllMocks()
    })

    test('Snapshot', () => {
        const tree = render(<BookCard id={1} url="abc.com" styleSelect="custom" hideIcon={true} />)
        act(() => {
            expect(tree).toMatchSnapshot()
        })
    })
    test('test with passing favoritebook variable', () => {
        const { getByTestId } = render(<BookCard />)

        act(() => {
            const state = { appReducer: { favorite: ['abc'] } }
            const isFocussed = jest.spyOn(redux, 'useSelector').mockImplementation((val) => {
                val(state)
                isFocussed.mockClear()
            })
        })
        // fireEvent.press(getByTestId('search'));
    })

    test('Fireevent heart', async () => {
        const { getByTestId } = render(<BookCard />)

        act(() => {
            fireEvent.press(getByTestId('heart'))
        })
    })
    test('Fireevent heart', async () => {
        const { getByTestId } = render(<BookCard />)

        act(() => {
            fireEvent.press(getByTestId('heart'))
            fireEvent.press(getByTestId('heart'))
        })
    })
})
