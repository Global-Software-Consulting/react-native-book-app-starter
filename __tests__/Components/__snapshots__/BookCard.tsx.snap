// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component testing Snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "transparent",
      "borderRadius": 20,
      "elevation": 2,
      "margin": 2,
      "marginRight": 3,
      "shadowColor": "red",
      "shadowOffset": Object {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 3,
      "shadowRadius": 2,
    }
  }
>
  <mockedDefaultExport
    resizeMode="contain"
    source={
      Object {
        "priority": undefined,
        "uri": "abc.com",
      }
    }
    style={
      Object {
        "alignSelf": "center",
        "borderRadius": 20,
        "height": 400,
        "margin": 5,
        "width": 487.5,
      }
    }
  />
  <Text
    style={
      Object {
        "alignSelf": "center",
        "fontSize": 37.5,
        "fontWeight": "500",
        "maxWidth": 450,
        "textAlign": "center",
      }
    }
  />
  <Text
    style={
      Object {
        "alignSelf": "center",
        "color": "red",
        "fontSize": 30,
        "fontWeight": "300",
        "maxWidth": 412.5,
      }
    }
  />
</View>
`;
